### PR TITLE
Adding new modules and data SpotFindingResults class. 

### PR DESCRIPTION
**Pull request by shanaxel42**
This PR lays some of the ground work for the spot finding refactor. 

- Adds two new modules FindSpots and DecodeSpots (which will be the only ones left by the end of this).
- Adds their base class api contracts. 
- Adds new class SpotFindingResults which is a wrapper class around a dictionary of round/ch index pairs and the spots found there (SpotAttributes).

Depends on #1489

### CODE REVIEW

File: starfish/core/types/_spot_finding_resutls.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            If spots were fond using Imagestack.transform() the result is a list of
+            tuples (indices, SpotAttributes). Instantiating SpotFindingResults with
+            this list will convert the information to a dictionary.
+        """
+        self._results: Mapping[Tuple, SpotAttributes] = dict()
+        if spot_attributes_list:
+            for indices, spots in spot_attributes_list:
+                self._results[indices] = spots
+
+    def set_spots_for_round_ch(self, indices: Mapping[Axes, int], spots: SpotAttributes
```
Comment (ttung) [inline starfish/core/types/_spot_finding_resutls.py]: nit: this function declaration style leads to unnecessary LOC changes.  would prefer:

```
    def set_spots_for_round_ch(self, indices: Mapping[Axes, int], spots: SpotAttributes) -> None:
```

or

```
    def set_spots_for_round_ch(
        self, indices: Mapping[Axes, int], spots: SpotAttributes) -> None:
```

or 

```
    def set_spots_for_round_ch(
        self, indices: Mapping[Axes, int], spots: SpotAttributes,
    ) -> None:
```

All of these approaches do not produce unnecessary changes if the method name changes.

File: starfish/core/types/_spot_finding_resutls.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+        Parameters
+        ----------
+        indices: Mapping[Axes, int]
+            Mapping of Axes to int values
+        spots: SpotAttributes
+            Describes spots found on this tile.
+        """
+        round_ch_index = tuple(indices[i] for i in AXES_ORDER)
+        self._results[round_ch_index] = spots  # type: ignore
```
Comment (ttung) [inline starfish/core/types/_spot_finding_resutls.py]: You can get rid of the `# type: ignore` if you make self._results a `MutableMapping`.

File: starfish/core/types/_spot_finding_resutls.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        """
+        self._results: Mapping[Tuple, SpotAttributes] = dict()
+        if spot_attributes_list:
+            for indices, spots in spot_attributes_list:
+                self._results[indices] = spots
+
+    def set_spots_for_round_ch(self, indices: Mapping[Axes, int], spots: SpotAttributes
+                               ) -> None:
+        """
+        Add the r,ch indices and corresponding SpotAttributes to the results dict.
```
Comment (ttung) [inline starfish/core/types/_spot_finding_resutls.py]:
```suggestion
        Add the round, ch indices and corresponding SpotAttributes to the results dict.
```

to match what you are doing elsewhere.

File: starfish/core/types/_spot_finding_resutls.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        Return all SpotAttributes across rounds and chs.
+        """
+        return self._results.values()
+
+    @property
+    def round_labels(self):
+        """
+        Return the set of Round labels in the SpotFindingResults
+        """
+        return list(set(sorted(r for (r, ch) in self.round_ch_indices())))
```
Comment (ttung) [inline starfish/core/types/_spot_finding_resutls.py]: if you take a set of sorted, then you lose the sorting.  did you want sorted(set(...))?

File: starfish/core/types/_spot_finding_resutls.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+    @property
+    def ch_labels(self):
+        """
+        Return the set of ch labels in the SpotFindingResults
+        """
+        return list(set(sorted(ch for (c, ch) in self.round_ch_indices())))
```
Comment (ttung) [inline starfish/core/types/_spot_finding_resutls.py]: ditto.

File: starfish/core/types/__init__.py [hunk 6]
```diff
@@ -15,6 +15,7 @@
 )
 from ._decoded_spots import DecodedSpots
 from ._spot_attributes import SpotAttributes
+from ._spot_finding_resutls import SpotFindingResults
```
Comment (ttung) [inline starfish/core/types/__init__.py]:
```suggestion
from ._spot_finding_results import SpotFindingResults
```

(and as such, the file probably needs to be renamed)

File: starfish/core/types/_spot_finding_resutls.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
+from typing import List, Mapping, MutableMapping, Optional, Tuple
+
+from starfish.core.types import Axes, SpotAttributes
+
+AXES_ORDER = (Axes.ROUND, Axes.CH)
+
+
+class SpotFindingResults:
```
Comment (ambrosejcarr) [inline starfish/core/types/_spot_finding_resutls.py]: Given this wraps a dict, I would prefer this be implemented to mimic the dictionary container, and more clearly document what it contains. This should simplify a lot of the accessors you've written. For example, 

```diff
+     def values(self):
-     def all_spots(self):
-         """
-         Return all SpotAttributes across rounds and chs.
-         """
          return self._results.values()
```

See: https://docs.python.org/2.5/ref/sequence-types.html

Comment (ttung) [inline starfish/core/types/_spot_finding_resutls.py]: Couple design questions:

1. Is it ever the case we would want to do spot finding on a zplane basis?
2. Why not just pass in round, ch as arguments to the methods in this class?

File: starfish/core/types/_spot_finding_resutls.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        Return the set of Round labels in the SpotFindingResults
+        """
+        return list(sorted(set(r for (r, ch) in self.round_ch_indices())))
+
+    @property
+    def ch_labels(self):
+        """
+        Return the set of ch labels in the SpotFindingResults
+        """
+        return list(sorted(set(ch for (c, ch) in self.round_ch_indices())))
```
Comment (ttung) [inline starfish/core/types/_spot_finding_resutls.py]:
```suggestion
        return list(sorted(set(ch for (r, ch) in self.round_ch_indices())))
```

File: starfish/core/types/_spot_finding_resutls.py (hunk 7 above)
Comment (shanaxel42) [reply starfish/core/types/_spot_finding_resutls.py]: Yea so I started down this route but found that things became too confusing. For example get_spots_for_round_ch would instead just be  __getitem__(key) and then when you used it it was totally clear what you were getting. I'm not super opinionated on this though so open to changing it back if that's what you'd prefer I just thought this was more descriptive

Comment (shanaxel42) [reply starfish/core/types/_spot_finding_resutls.py]: I think tony and I resolved these in person, but tony yell at me if not

File: starfish/core/types/_spot_finding_results.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+    def __init__(self, spot_attributes_list: Optional[List[Tuple]] = None):
+        """
+        Construct a SpotFindingResults instance
+
+        Parameters
+        -----------
+        spot_attributes_list : Optional[List[Tuple]]
+            If spots were found using ImageStack.transform() the result is a list of
+            tuples (indices, SpotAttributes). Instantiating SpotFindingResults with
```
Comment (ttung) [inline starfish/core/types/_spot_finding_results.py]: The format for the indices is not very clear.  It appears to be a Tuple of index values in the order of AXES_ORDER, but I'm just inferring that from the code.

File: starfish/core/types/_spot_finding_results.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        Return all SpotAttributes across rounds and chs.
+        """
+        return self._results.values()
+
+    @property
+    def round_labels(self):
+        """
+        Return the set of Round labels in the SpotFindingResults
+        """
+        return list(sorted(set(r for (r, ch) in self.keys())))
```
Comment (ttung) [inline starfish/core/types/_spot_finding_results.py]: i still think sorted(set(..)) would do what you want.  the extra list() shouldn't be necessary.

```
>>> sorted(set([1, 2, 3, 4]))
[1, 2, 3, 4]
>>> ^D
```

File: starfish/core/types/_spot_finding_results.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    def values(self):
+        """
+        Return all SpotAttributes across rounds and chs.
+        """
+        return self._results.values()
+
+    @property
+    def round_labels(self):
+        """
+        Return the set of Round labels in the SpotFindingResults
```
Comment (ttung) [inline starfish/core/types/_spot_finding_results.py]:
```suggestion
        Return the set of round labels in the SpotFindingResults
```

File: starfish/core/types/_spot_finding_results.py (hunk 9 above)
Comment (shanaxel42) [reply starfish/core/types/_spot_finding_results.py]: That's exactly what it is! I'll make it more clear

File: docs/source/api/spots/index.rst [hunk 12]
```diff
@@ -38,6 +38,18 @@ Spot Detectors can be imported using ``starfish.spots.DetectSpots``, which regis
...
+NOTE: Starfish is embarking on a SpotFinding data structures refactor see `Spot Finding Refactor Plan`_
+DetectSpots will be replaced by the FindSpots module documented below.
+
+.. _`Spot Finding Refactor Plan`: https://github.com/spacetx/starfish/issues/1514
+
+.. code-block:: python
+
+    from starfish.spots import FindSpots
+
+.. autoclass:: starfish.spots.FindSpots
```
Comment (ttung) [inline docs/source/api/spots/index.rst]: Isn't FindSpots a module?  I think you want `automodule` here to recursively pull in all the implementations in `FindSpots`, and the weird logic to import in all the submodules of `FindSpots`.

File: docs/source/api/spots/index.rst [hunk 13]
```diff
@@ -53,6 +65,15 @@ Decoders can be imported using ``starfish.spots.Decode``, which registers all cl
 .. autoclass:: starfish.spots.Decode
    :members:
 
+NOTE: Starfish is embarking on a SpotFinding data structures refactor see `Spot Finding Refactor Plan`_
+
+.. code-block:: python
+
+    from starfish.spots import DecodeSpots
```
Comment (ttung) [inline docs/source/api/spots/index.rst]: Ditto.

File: starfish/core/spots/FindSpots/_base.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+from starfish.core.pipeline.algorithmbase import AlgorithmBase
+from starfish.core.types import Number, SpotFindingResults
+
+
+class FindSpotsAlgorithm(metaclass=AlgorithmBase):
+
+    @abstractmethod
+    def run(self, image_stack: ImageStack,
+            reference_image: Optional[ImageStack] = None, *args) -> SpotFindingResults:
+        """Find and measure spots across rounds and channels in the provided
```
Comment (ttung) [inline starfish/core/spots/FindSpots/_base.py]: unwrap?  I'd check your editor to make sure you're set up to 100col.  This clocks in at 92col unwrapped for me.
